# HW6 Comprehensive Supervised Learning
Contributors: Zongyu Wu, Tony Wilson, Chandler Smith

Summary: The overall purpose of this assignment is to tie together a variety of supervised learning techniques in order to appropriately analyze several questions related to the Behavioral Risk Factor Surveillance System. More specifically, the goal is to use supervised ML to identify patterns of comorbidity among the survey respondents. This is a comparative exercise focusing on pre (2019) and post (2021) covid health. 

Notes from discussion:

Approach: One approach for understanding behavioral factors behind comorbidity is to apply classification algorithms to the BRFSS data. Such algorithms can be trained to predict whether an individual has multiple chronic conditions based on their responses to the survey questions. By analyzing the features that are most important for predicting comorbidity, one can identify risk factors and inform the development of targeted prevention and intervention strategies.

Response related:
Depression = ADDEPEV3
 Ever told Asthma = _CASTHM1
 COPD = (CHCCOPD2 for 2019 and CHCCOPD3 for 2021)
 Cancer = (combine CHCSCNCR and CHCOCNCR)
 Ever told Heart Condition = combination of CVDCRHD4, CVDINFR4 and CVDSTRK3
 Diabetes = DIABETE4
 

Key demographic features:
Age: _AGE_G
 Marital: MARITAL
 Sex: _SEX
 Income: INCOME2
 Education: _EDUCAG

- Do NOT refer to correlation
- BRFSS data can be used to understand comorbidity, which refers to the presence of multiple chronic conditions in an individual. 
- RFSS uses a complex sampling and weighting scheme to measure prevalence of many health conditions, behavioral and lifestyle related risk factors and emerging health issues in states. 
- Do not use the weight variable as we are conducting estimations. 


- 2019 Codebook: https://www.cdc.gov/brfss/annual_data/2019/pdf/codebook19_llcp-v2-508.HTML
- 2021 Codebook: https://www.cdc.gov/brfss/annual_data/2021/pdf/codebook21_llcp-v2-508.pdf

## Clean, Standardize, and Merge data


In [6]:
import pandas as pd
import numpy as np
import os
from datetime import datetime

import sklearn 

# Libraries related to outlier detection
from sklearn.neighbors import LocalOutlierFactor
from sklearn.ensemble import IsolationForest
from sklearn.covariance import EllipticEnvelope

from sklearn import datasets
import random
import seaborn as sns
import matplotlib.pyplot as plt
import missingno as msno
from scipy.stats import pearsonr

import warnings
warnings.filterwarnings('ignore') 
sns.set(rc={'figure.figsize':(11,8)})

pd.options.display.float_format = '{:.2f}'.format

## <font color= darkgreen> Basic EDA

### <font color= lightblue> Step 1: Read the data and merge into Pandas Data Frame

In [7]:
###############################################################################################################################################################################
'''                                                                       Data Frame Setup                                                                                  '''
###############################################################################################################################################################################
## The first column is index: skipping that column to end read csv to Panda Data Frame
df_21 = pd.read_csv("res/brfss21-1.csv")
df_21.drop(columns="Unnamed: 0", inplace=True)
df_21.head()



_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE  \
0    1.00    1.00  b'01192021'  b'01'  b'19'  b'2021'   1100.00   
1    1.00    1.00  b'01212021'  b'01'  b'21'  b'2021'   1100.00   
2    1.00    1.00  b'01212021'  b'01'  b'21'  b'2021'   1100.00   
3    1.00    1.00  b'01172021'  b'01'  b'17'  b'2021'   1100.00   
4    1.00    1.00  b'01152021'  b'01'  b'15'  b'2021'   1100.00   

           SEQNO          _PSU  CTELENM1  ...  _FRTRES1  _VEGRES1  _FRUTSU1  \
0  b'2021000001' 2021000001.00      1.00  ...      1.00      1.00    100.00   
1  b'2021000002' 2021000002.00      1.00  ...      1.00      1.00    100.00   
2  b'2021000003' 2021000003.00      1.00  ...      1.00      1.00    100.00   
3  b'2021000004' 2021000004.00      1.00  ...      1.00      1.00    114.00   
4  b'2021000005' 2021000005.00      1.00  ...      1.00      1.00    100.00   

   _VEGESU1  _FRTLT1A  _VEGLT1A  _FRT16A  _VEG23A  _FRUITE1  _VEGETE1  
0    214.00      1.00      1.00     1.00     1.00      0.00      0.00  
1    128.00      1.00      1.00     1.00     1.00      0.00      0.00  
2     71.00      1.00      2.00     1.00     1.00      0.00      0.00  
3    165.00      1.00      1.00     1.00     1.00      0.00      0.00  
4    258.00      1.00      1.00     1.00     1.00      0.00      0.00  

[5 rows x 250 columns]

In [8]:
df_19 = pd.read_csv("res/brfss19-1.csv")
df_19.drop(columns="Unnamed: 0", inplace= True)
df_19.head()


_STATE  FMONTH    IDATE  IMONTH  IDAY  IYEAR  DISPCODE       SEQNO  \
0       1       1  1182019       1    18   2019      1100  2019000001   
1       1       1  1132019       1    13   2019      1100  2019000002   
2       1       1  1182019       1    18   2019      1100  2019000003   
3       1       1  1182019       1    18   2019      1200  2019000004   
4       1       1  1042019       1     4   2019      1100  2019000005   

         _PSU  CTELENM1  ...  _VEGESU1  _FRTLT1A  _VEGLT1A  _FRT16A  _VEG23A  \
0  2019000001      1.00  ...    114.00         1         1        1        1   
1  2019000002      1.00  ...    121.00         1         1        1        1   
2  2019000003      1.00  ...    164.00         1         1        1        1   
3  2019000004      1.00  ...       NaN         9         9        1        1   
4  2019000005      1.00  ...    178.00         1         1        1        1   

   _FRUITE1  _VEGETE1  _FLSHOT7  _PNEUMO3  _AIDTST4  
0         0         0      2.00      1.00      2.00  
1         0         0      1.00      1.00      2.00  
2         0         0      1.00      2.00      2.00  
3         1         1      9.00      9.00       NaN  
4         0         0      2.00      1.00      2.00  

[5 rows x 250 columns]

### <font color= lightblue> Step 2: Summary of Stats

In [9]:
###############################################################################################################################################################################
'''                                                                    Data Frame Description                                                                               '''
###############################################################################################################################################################################
df_19.describe()

_STATE    FMONTH       IDATE    IMONTH      IDAY     IYEAR  DISPCODE  \
count 418268.00 418268.00   418268.00 418268.00 418268.00 418268.00 418268.00   
mean      29.62      6.54  6727352.51      6.58     14.53   2019.04   1117.44   
std       16.15      3.34  3304672.99      3.31      8.49      0.21     37.95   
min        1.00      1.00  1012020.00      1.00      1.00   2019.00   1100.00   
25%       18.00      4.00  4082019.00      4.00      7.00   2019.00   1100.00   
50%       27.00      7.00  7012019.00      7.00     14.00   2019.00   1100.00   
75%       42.00      9.00  9302019.00      9.00     22.00   2019.00   1100.00   
max       72.00     12.00 12312019.00     12.00     31.00   2020.00   1200.00   

              SEQNO          _PSU  CTELENM1  ...  _VEGESU1  _FRTLT1A  \
count     418268.00     418268.00 149941.00  ... 364838.00 418268.00   
mean  2019004884.52 2019004884.52      1.00  ...    204.14      2.19   
std         3653.32       3653.32      0.00  ...    267.90      2.40   
min   2019000001.00 2019000001.00      1.00  ...      0.00      1.00   
25%   2019002011.00 2019002011.00      1.00  ...    114.00      1.00   
50%   2019004137.00 2019004137.00      1.00  ...    165.00      1.00   
75%   2019006895.00 2019006895.00      1.00  ...    229.00      2.00   
max   2019017419.00 2019017419.00      1.00  ...  13204.00      9.00   

       _VEGLT1A   _FRT16A   _VEG23A  _FRUITE1  _VEGETE1  _FLSHOT7  _PNEUMO3  \
count 418268.00 418268.00 418268.00 418268.00 418268.00 159112.00 159112.00   
mean       2.19      1.00      1.00      0.11      0.13      2.23      2.37   
std        2.63      0.04      0.05      0.32      0.35      2.47      2.73   
min        1.00      0.00      0.00      0.00      0.00      1.00      1.00   
25%        1.00      1.00      1.00      0.00      0.00      1.00      1.00   
50%        1.00      1.00      1.00      0.00      0.00      1.00      1.00   
75%        2.00      1.00      1.00      0.00      0.00      2.00      2.00   
max        9.00      1.00      1.00      2.00      2.00      9.00      9.00   

       _AIDTST4  
count 377977.00  
mean       1.97  
std        1.56  
min        1.00  
25%        1.00  
50%        2.00  
75%        2.00  
max        9.00  

[8 rows x 250 columns]

In [10]:
###############################################################################################################################################################################
'''                                                                     Data Frame Shape                                                                                    '''
###############################################################################################################################################################################
df_19.shape

(418268, 250)

In [11]:
###############################################################################################################################################################################
'''                                                                    Data Frame Description                                                                               '''
###############################################################################################################################################################################
df_21.describe()

_STATE    FMONTH  DISPCODE          _PSU  CTELENM1  PVTRESD1  \
count 438693.00 438693.00 438693.00     438693.00 117786.00 117786.00   
mean      30.74      6.41   1118.19 2021006064.89      1.00      1.00   
std       15.33      3.42     38.58       6383.75      0.00      0.02   
min        1.00      1.00   1100.00 2021000001.00      1.00      1.00   
25%       20.00      3.00   1100.00 2021002091.00      1.00      1.00   
50%       31.00      6.00   1100.00 2021004338.00      1.00      1.00   
75%       41.00      9.00   1100.00 2021007674.00      1.00      1.00   
max       78.00     12.00   1200.00 2021039095.00      2.00      2.00   

       COLGHOUS  STATERE1   LADULT1  COLGSEX  ...  _FRTRES1  _VEGRES1  \
count     30.00 117786.00 117786.00    30.00  ... 438693.00 438693.00   
mean       1.00      1.00      1.01     1.63  ...      0.88      0.86   
std        0.00      0.00      0.08     0.49  ...      0.32      0.34   
min        1.00      1.00      1.00     1.00  ...      0.00      0.00   
25%        1.00      1.00      1.00     1.00  ...      1.00      1.00   
50%        1.00      1.00      1.00     2.00  ...      1.00      1.00   
75%        1.00      1.00      1.00     2.00  ...      1.00      1.00   
max        1.00      1.00      2.00     2.00  ...      1.00      1.00   

       _FRUTSU1  _VEGESU1  _FRTLT1A  _VEGLT1A   _FRT16A   _VEG23A  _FRUITE1  \
count 387606.00 378566.00 438693.00 438693.00 438693.00 438693.00 438693.00   
mean     178.34    271.54      2.27      2.26      0.99      0.99      0.13   
std      691.29   1036.23      2.49      2.71      0.07      0.09      0.35   
min        0.00      0.00      1.00      1.00      0.00      0.00      0.00   
25%       57.00    114.00      1.00      1.00      1.00      1.00      0.00   
50%      100.00    167.00      1.00      1.00      1.00      1.00      0.00   
75%      200.00    229.00      2.00      2.00      1.00      1.00      0.00   
max    19800.00  39600.00      9.00      9.00      1.00      1.00      2.00   

       _VEGETE1  
count 438693.00  
mean       0.15  
std        0.38  
min        0.00  
25%        0.00  
50%        0.00  
75%        0.00  
max        2.00  

[8 rows x 245 columns]

In [12]:
###############################################################################################################################################################################
'''                                                                     Data Frame Shape                                                                                    '''
###############################################################################################################################################################################
df_21.shape

(438693, 250)

### <font color= lightblue> Step 3: Choose Feature Space

<font color=white>We will first need to choose our feature space.  The five demographics provided will be the first on the list:
<font color= red>
* Age:  _AGE_G
* Marital: MARITAL
* Sex: _SEX
* Income: INCOME2
* Education: _EDUCAG

<font color= white>
Next we want to choose some features which we think will have some bearing on the response columns. We also want to pair the 250 features down to 20 (as per professor). Another demographic could be their race and city vs rural living as well as access to health care (insurance) and routine checkups:
<br><br>
<font color= red>

* Race: _RACE
* Urban / Rural: _METSTAT
* Health Care Access (Insurance): PRIMINSR &ensp;&ensp;&ensp;&ensp;<font color=lightblue>-- Health Insurance data not included in the data provided<font color= red>
* Health Care Access (Routine checkup): CHECKUP1

<font color= white>
Also, we should also track physical activity, High blood pressure, High Cholesterol, smoking habits, drinking habits, BMI, kidney disease
<br><br>
<font color= red>

* Physical exersice: _TOTINDA
* HBP: _RFHYPE6 &ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;<font color=lightblue>-- Blood Pressure metrics not included in the data provided<font color= white><br>

Instead we will use Currently taking blood pressure medication and Be very carefull with our clean up <br><font color= red>

* High Cholesterol: _RFCHOL3 &ensp;&ensp;&ensp;&ensp;<font color=lightblue>-- Cholesterol metrics not included in the data provided<font color= red>
* Four level smoker status: _SMOKER3
* Number of drinks per week: _DRNKWK1&ensp;&ensp;&ensp;&ensp;<font color=lightblue> Continuous Data<font color= red>
* Total Fruit per Day: _FRUTSU1&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;<font color=lightblue> Continuous Data<font color= red>
* Total Vegetables per Day: _VEGESU1&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;<font color=lightblue> Continuous Data<font color= red>
* Total French Fry per Day: FRNCHDA_&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;<font color=lightblue> Continuous Data<font color= red>
* Body Mass Index: _BMI5&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;<font color=lightblue> Continuous Data<font color= red>
* Kidney Disease: CHCKDNY2

<font color= white>
The last demographic that Tony would like to include specifically is whether the participant is a veteran:
br><br>
<font color= red>

* Veteran Status: VETERAN3


In [13]:
###############################################################################################################################################################################
'''                                                               Choosing Response and Features                                                                            '''
###############################################################################################################################################################################
df_19_trim = df_19.loc[:, ['ADDEPEV3', '_CASTHM1', 'CHCCOPD2', 'CHCSCNCR', 'CHCOCNCR', 'CVDCRHD4', 'CVDINFR4', 'CVDSTRK3',
                            'DIABETE4', '_AGE_G', 'MARITAL', '_SEX', 'INCOME2', '_EDUCAG', '_RACE', '_METSTAT', 'CHECKUP1',
                            'BPMEDS', '_TOTINDA', '_SMOKER3', '_DRNKWK1', '_FRUTSU1', '_VEGESU1','FRNCHDA_', '_BMI5', 
                            'CHCKDNY2', 'VETERAN3']]
df_19_trim.shape

(418268, 27)

In [14]:
###############################################################################################################################################################################
'''                                                               Choosing Response and Features                                                                            '''
###############################################################################################################################################################################
df_21_trim = df_21.loc[:, ['ADDEPEV3', '_CASTHM1', 'CHCCOPD3', 'CHCSCNCR', 'CHCOCNCR', 'CVDCRHD4', 'CVDINFR4', 'CVDSTRK3',
                            'DIABETE4', '_AGE_G', 'MARITAL', '_SEX', 'INCOME3', '_EDUCAG', '_RACE', '_METSTAT', 'CHECKUP1',
                            'BPMEDS', '_TOTINDA', '_SMOKER3', '_DRNKWK1', '_FRUTSU1', '_VEGESU1','FRNCHDA_', '_BMI5', 
                            'CHCKDNY2', 'VETERAN3']]
df_21_trim.shape

(438693, 27)

### <font color= lightblue> Step 4: Missing Data Analysis

##### BPMEDS
From looking at the data itself, most of the missing values from BPMEDS is because the previous question answered something other than yes to every choice other than yes.  Therefore we can infer that these were a no and that 7 and 9 should be dropped.
* 2021-
- missing: 266,560 
<br>vs<br>
- Previous other than yes: 266,560

there is only 0.26% which was I don't know or Refused

* 2019-
- missing: 248,634 
<br>vs<br>
- Previous other than yes: 248,634

there is only 0.21% which was I don't know or Refused

In [15]:
###############################################################################################################################################################################
'''                                                                   Analyze, Clean, and Impute                                                                            '''
###############################################################################################################################################################################
## Print Original Shape for comparisson
print('Original 19 shape:\t', df_21_trim.shape)
print('Original 21 shape:\t', df_21_trim.shape)

## Interpret and clean df
# output how many NaN there is per column
print('Missing values by field (Pre-Interpolate):')

# Visualize the finished Data Frame
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(df_19_trim.isnull().sum(), '\n')
    print(df_21_trim.isnull().sum(), '\n')

## Replace missing values in BPMEDS with No answer and drop 7 and 9 response as NAN
df_19_trim = df_19_trim[df_19_trim['BPMEDS'] != 7] 
df_19_trim = df_19_trim[df_19_trim['BPMEDS'] != 9]
df_19_trim['BPMEDS'].fillna(2, inplace= True) 

# How are we doing now
print(df_19_trim.isnull().sum(), '\n')
print(df_19_trim.shape)

# Drop the remaining NaN
df_19_trim.dropna(inplace= True)


## Replace missing values in BPMEDS with No answer and drop 7 and 9 response as NAN
df_21_trim = df_21_trim[df_21_trim['BPMEDS'] != 7] 
df_21_trim = df_21_trim[df_21_trim['BPMEDS'] != 9]
df_21_trim['BPMEDS'].fillna(2, inplace= True) 

# How are we doing now
print(df_21_trim.isnull().sum(), '\n')
print(df_21_trim.shape)

# Drop the remaining NaN
df_21_trim.dropna(inplace= True)

# Output final shape
print('Final 19 shape:\t', df_19_trim.shape)
print('Final 21 shape:\t', df_21_trim.shape)

Original 19 shape:	 (438693, 27)
Original 21 shape:	 (438693, 27)
Missing values by field (Pre-Interpolate):
ADDEPEV3        10
_CASTHM1         0
CHCCOPD2         8
CHCSCNCR         8
CHCOCNCR         9
CVDCRHD4         8
CVDINFR4        10
CVDSTRK3        11
DIABETE4         9
_AGE_G           0
MARITAL         49
_SEX             0
INCOME2       6881
_EDUCAG          0
_RACE            3
_METSTAT      8458
CHECKUP1        10
BPMEDS      248634
_TOTINDA         0
_SMOKER3         0
_DRNKWK1         0
_FRUTSU1     44600
_VEGESU1     53430
FRNCHDA_     38866
_BMI5        36203
CHCKDNY2        11
VETERAN3      1374
dtype: int64 

ADDEPEV3         3
_CASTHM1         0
CHCCOPD3         3
CHCSCNCR         2
CHCOCNCR         3
CVDCRHD4         2
CVDINFR4         2
CVDSTRK3         2
DIABETE4         3
_AGE_G           0
MARITAL          5
_SEX             0
INCOME3       8847
_EDUCAG          0
_RACE            0
_METSTAT      7054
CHECKUP1         2
BPMEDS      266560
_TOTINDA         0
_S

## Create categorical columns

- Column 1: Binary - indicates whether an individual has haf any of the chronic conditions. 
- Column 2: Multiclass - create a milticlass column with values that inidicate the total number of chronic conditions.

### <font color= lightblue> Step 5: Binary and Multiclass Response Created

<font color= yellow>First we create the binaries for 2019

In [16]:
###############################################################################################################################################################################
'''                                                                   Convert Response Column for 2019                                                                      '''
###############################################################################################################################################################################
## Binary Comorbidity Response Features from our respective responses
# Create the Depression comorbidity variable
df_19_trim['Depression'] = df_19_trim.apply(lambda row: 1 if row[0] == 1 else (0 if row[0] == 2 else np.NAN), axis=1)
print(df_19_trim[['Depression', 'ADDEPEV3']].value_counts())
print(df_19_trim['Depression'].isnull().sum())
print(df_19_trim['ADDEPEV3'].value_counts(), '\n')

# Create the Asthma comorbidity variable
df_19_trim['Asthma'] = df_19_trim.apply(lambda row: 1 if row[1] == 1 else (0 if row[1] == 2 else np.NAN), axis=1)
print(df_19_trim[['Asthma', '_CASTHM1']].value_counts())
print(df_19_trim['Asthma'].isnull().sum())
print(df_19_trim['_CASTHM1'].value_counts(), '\n')

# Create the COPD comorbidity variable
df_19_trim['COPD'] = df_19_trim.apply(lambda row: 1 if row[2] == 1 else (0 if row[2] == 2 else np.NAN), axis=1)
print(df_19_trim[['COPD', 'CHCCOPD2']].value_counts())
print(df_19_trim['COPD'].isnull().sum())
print(df_19_trim['CHCCOPD2'].value_counts(), '\n')

# Create the Cancer comorbidity variable
df_19_trim['Cancer'] = df_19_trim.apply(lambda row: 1 if row[3] == 1 else (1 if row[4] == 1 else (0 if row[3] == 2 else ( 0 if row[4] == 2 else np.NAN))), axis=1)
print(df_19_trim[['Cancer', 'CHCSCNCR', 'CHCOCNCR']].value_counts())
print(df_19_trim['Cancer'].isnull().sum())
print(df_19_trim['CHCSCNCR'].value_counts(), '\n')
print(df_19_trim['CHCOCNCR'].value_counts(), '\n')
print(df_19_trim['Cancer'].value_counts(), '\n')

# Create the Heart Condition comorbidity variable
df_19_trim['Heart'] = df_19_trim.apply(lambda row: 1 if row[5] == 1 else (1 if row[6] == 1 else (1 if row[7] == 1 else (0 if row[5] == 2 else (0 if row[6] == 2 else (0 if row[7] == 2 else np.NAN))))), axis=1)
print(df_19_trim[['Heart', 'CVDCRHD4', 'CVDINFR4', 'CVDSTRK3']].value_counts())
print(df_19_trim['Heart'].isnull().sum())
print(df_19_trim['CVDCRHD4'].value_counts(), '\n')
print(df_19_trim['CVDINFR4'].value_counts(), '\n')
print(df_19_trim['CVDSTRK3'].value_counts(), '\n')
print(df_19_trim['Heart'].value_counts(), '\n')

# Create the Depression comorbidity variable
df_19_trim['Diabetes'] = df_19_trim.apply(lambda row: 1 if row[8] == 1 else (0 if row[8] == 2 else (0 if row[8] == 3 else (0 if row[8] == 4 else np.NAN))), axis=1)
print(df_19_trim[['Diabetes', 'DIABETE4']].value_counts())
print(df_19_trim['Diabetes'].isnull().sum())
print(df_19_trim['DIABETE4'].value_counts(), '\n')

print(df_19_trim.head())

Depression  ADDEPEV3
0.00        2.00        264512
1.00        1.00         64620
dtype: int64
1373
2.00    264512
1.00     64620
7.00      1188
9.00       185
Name: ADDEPEV3, dtype: int64 

Asthma  _CASTHM1
1.00    1           297376
0.00    2            30814
dtype: int64
2315
1    297376
2     30814
9      2315
Name: _CASTHM1, dtype: int64 

COPD  CHCCOPD2
0.00  2.00        300875
1.00  1.00         28179
dtype: int64
1451
2.00    300875
1.00     28179
7.00      1398
9.00        53
Name: CHCCOPD2, dtype: int64 

Cancer  CHCSCNCR  CHCOCNCR
0.00    2.00      2.00        267482
1.00    1.00      2.00         26841
        2.00      1.00         26565
        1.00      1.00          8057
0.00    7.00      2.00           660
        2.00      7.00           509
1.00    7.00      1.00           143
        1.00      7.00            89
0.00    2.00      9.00            58
        9.00      2.00            13
1.00    1.00      9.00            10
        9.00      1.00             1
dtype: 

<font color= yellow>Then we create the binaries for 2021

In [17]:
###############################################################################################################################################################################
'''                                                                   Convert Response Column for 2021                                                                      '''
###############################################################################################################################################################################
## Binary Comorbidity Response Features from our respective responses
# Create the Depression comorbidity variable
df_21_trim['Depression'] = df_21_trim.apply(lambda row: 1 if row[0] == 1 else (0 if row[0] == 2 else np.NAN), axis=1)
print(df_21_trim[['Depression', 'ADDEPEV3']].value_counts())
print(df_21_trim['Depression'].isnull().sum())
print(df_21_trim['ADDEPEV3'].value_counts(), '\n')

# Create the Asthma comorbidity variable
df_21_trim['Asthma'] = df_21_trim.apply(lambda row: 1 if row[1] == 1 else (0 if row[1] == 2 else np.NAN), axis=1)
print(df_21_trim[['Asthma', '_CASTHM1']].value_counts())
print(df_21_trim['Asthma'].isnull().sum())
print(df_21_trim['_CASTHM1'].value_counts(), '\n')

# Create the COPD comorbidity variable
df_21_trim['COPD'] = df_21_trim.apply(lambda row: 1 if row[2] == 1 else (0 if row[2] == 2 else np.NAN), axis=1)
print(df_21_trim[['COPD', 'CHCCOPD3']].value_counts())
print(df_21_trim['COPD'].isnull().sum())
print(df_21_trim['CHCCOPD3'].value_counts(), '\n')

# Create the Cancer comorbidity variable
df_21_trim['Cancer'] = df_21_trim.apply(lambda row: 1 if row[3] == 1 else (1 if row[4] == 1 else (0 if row[3] == 2 else ( 0 if row[4] == 2 else np.NAN))), axis=1)
print(df_21_trim[['Cancer', 'CHCSCNCR', 'CHCOCNCR']].value_counts())
print(df_21_trim['Cancer'].isnull().sum())
print(df_21_trim['CHCSCNCR'].value_counts(), '\n')
print(df_21_trim['CHCOCNCR'].value_counts(), '\n')
print(df_21_trim['Cancer'].value_counts(), '\n')

# Create the Heart Condition comorbidity variable
df_21_trim['Heart'] = df_21_trim.apply(lambda row: 1 if row[5] == 1 else (1 if row[6] == 1 else (1 if row[7] == 1 else (0 if row[5] == 2 else (0 if row[6] == 2 else (0 if row[7] == 2 else np.NAN))))), axis=1)
print(df_21_trim[['Heart', 'CVDCRHD4', 'CVDINFR4', 'CVDSTRK3']].value_counts())
print(df_21_trim['Heart'].isnull().sum())
print(df_21_trim['CVDCRHD4'].value_counts(), '\n')
print(df_21_trim['CVDINFR4'].value_counts(), '\n')
print(df_21_trim['CVDSTRK3'].value_counts(), '\n')
print(df_21_trim['Heart'].value_counts(), '\n')

# Create the Depression comorbidity variable
df_21_trim['Diabetes'] = df_21_trim.apply(lambda row: 1 if row[8] == 1 else (0 if row[8] == 2 else (0 if row[8] == 3 else (0 if row[8] == 4 else np.NAN))), axis=1)
print(df_21_trim[['Diabetes', 'DIABETE4']].value_counts())
print(df_21_trim['Diabetes'].isnull().sum())
print(df_21_trim['DIABETE4'].value_counts(), '\n')

print(df_21_trim.head())

Depression  ADDEPEV3
0.00        2.00        268786
1.00        1.00         68765
dtype: int64
1447
2.00    268786
1.00     68765
7.00      1183
9.00       264
Name: ADDEPEV3, dtype: int64 

Asthma  _CASTHM1
1.00    1.00        303552
0.00    2.00         33089
dtype: int64
2357
1.00    303552
2.00     33089
9.00      2357
Name: _CASTHM1, dtype: int64 

COPD  CHCCOPD3
0.00  2.00        311016
1.00  1.00         26759
dtype: int64
1223
2.00    311016
1.00     26759
7.00      1157
9.00        66
Name: CHCCOPD3, dtype: int64 

Cancer  CHCSCNCR  CHCOCNCR
0.00    2.00      2.00        278474
1.00    2.00      1.00         25632
        1.00      2.00         25540
                  1.00          7869
0.00    7.00      2.00           588
        2.00      7.00           468
1.00    7.00      1.00           137
        1.00      7.00            88
0.00    2.00      9.00            56
        9.00      2.00            17
1.00    1.00      9.00            14
        9.00      1.00             

<font color= yellow>Drop the coloumns no longer needed and any missing values introduced

In [18]:
###############################################################################################################################################################################
'''                                                                   Re-Clean Data fropm both years                                                                        '''
###############################################################################################################################################################################
## First we need to drop the columns we no longer need and then the new NaNs for 2019
print('df_19_trim Original Shape_\t', df_19_trim.shape)
df_19_trim.drop(columns=['ADDEPEV3', '_CASTHM1', 'CHCCOPD2', 'CHCSCNCR', 'CHCOCNCR', 'CVDCRHD4', 'CVDINFR4', 'CVDSTRK3', 'DIABETE4'], inplace=True)
df_19_trim.dropna(inplace= True)
print('df_19_trim New Shape_\t', df_19_trim.shape, '\n')

## First we need to drop the columns we no longer need and then the new NaNs for 2021
print('df_21_trim Original Shape_\t', df_21_trim.shape)
df_21_trim.drop(columns=['ADDEPEV3', '_CASTHM1', 'CHCCOPD3', 'CHCSCNCR', 'CHCOCNCR', 'CVDCRHD4', 'CVDINFR4', 'CVDSTRK3', 'DIABETE4'], inplace=True)
df_21_trim.dropna(inplace= True)
print('df_19_trim New Shape_\t', df_21_trim.shape)

df_19_trim Original Shape_	 (330505, 33)
df_19_trim New Shape_	 (325245, 24) 

df_21_trim Original Shape_	 (338998, 33)
df_19_trim New Shape_	 (333899, 24)


<font color= yellow>Now we can move on to our last Comorbidity Variable for each year

In [19]:
###############################################################################################################################################################################
'''                                                                   Convert Response Column for 2019/2021                                                                 '''
###############################################################################################################################################################################
## Multiclass Response from the Binary Comorbidities
# First for 2019
df_19_trim['Comorbidity'] = df_19_trim.apply(lambda row: 1 if row[18] == 1 else (1 if row[19] == 1 else (1 if row[20] == 1 else (1 if row[21] == 1 else (1 if row[22] == 1 else (1 if row[23] == 1 else 0))))), axis=1)

# Then for 2021
df_21_trim['Comorbidity'] = df_21_trim.apply(lambda row: 1 if row[18] == 1 else (1 if row[19] == 1 else (1 if row[20] == 1 else (1 if row[21] == 1 else (1 if row[22] == 1 else (1 if row[23] == 1 else 0))))), axis=1)

### <font color= lightblue> Step 6: Make sure there are no NaNs left over

In [20]:
###############################################################################################################################################################################
'''                                                            Check for missing values again for 2019/2021                                                                 '''
###############################################################################################################################################################################
# output how many NaN there is per column
print('Missing values by field (Pre-Interpolate):')

# Visualize the finished Data Frame
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(df_19_trim.isnull().sum(), '\n')
    print(df_21_trim.isnull().sum(), '\n')

# We can see that there is one column left in each that have a different name so let us fix this now
df_21_trim = df_21_trim.rename(columns={'INCOME3': 'INCOME2'})

# Recheck by visualizing the finished Data Frame
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(df_19_trim.isnull().sum(), '\n')
    print(df_21_trim.isnull().sum(), '\n')

# Final shape of DF's
print('Final Shapes:\n', df_19_trim.shape, '\n', df_21_trim.shape)

Missing values by field (Pre-Interpolate):
_AGE_G         0
MARITAL        0
_SEX           0
INCOME2        0
_EDUCAG        0
_RACE          0
_METSTAT       0
CHECKUP1       0
BPMEDS         0
_TOTINDA       0
_SMOKER3       0
_DRNKWK1       0
_FRUTSU1       0
_VEGESU1       0
FRNCHDA_       0
_BMI5          0
CHCKDNY2       0
VETERAN3       0
Depression     0
Asthma         0
COPD           0
Cancer         0
Heart          0
Diabetes       0
Comorbidity    0
dtype: int64 

_AGE_G         0
MARITAL        0
_SEX           0
INCOME3        0
_EDUCAG        0
_RACE          0
_METSTAT       0
CHECKUP1       0
BPMEDS         0
_TOTINDA       0
_SMOKER3       0
_DRNKWK1       0
_FRUTSU1       0
_VEGESU1       0
FRNCHDA_       0
_BMI5          0
CHCKDNY2       0
VETERAN3       0
Depression     0
Asthma         0
COPD           0
Cancer         0
Heart          0
Diabetes       0
Comorbidity    0
dtype: int64 

_AGE_G         0
MARITAL        0
_SEX           0
INCOME2        0
_EDUCAG  

### <font color= lightblue> Step 6: Merge Data Frames

In [21]:
df = pd.concat([df_19_trim, df_21_trim])
df.head()

_AGE_G  MARITAL  _SEX  INCOME2  _EDUCAG  _RACE  _METSTAT  CHECKUP1  BPMEDS  \
0    6.00     2.00  2.00     3.00     1.00   2.00      1.00      1.00    1.00   
1    6.00     1.00  2.00     5.00     3.00   1.00      1.00      1.00    2.00   
2    6.00     3.00  2.00     7.00     4.00   2.00      1.00      1.00    1.00   
4    6.00     1.00  2.00    99.00     3.00   1.00      2.00      1.00    2.00   
6    6.00     2.00  1.00     7.00     4.00   1.00      2.00      1.00    2.00   

   _TOTINDA  ...   _BMI5  CHCKDNY2  VETERAN3  Depression  Asthma  COPD  \
0      2.00  ... 2817.00      2.00      2.00        0.00    1.00  0.00   
1      1.00  ... 1854.00      2.00      2.00        0.00    1.00  0.00   
2      1.00  ... 3162.00      2.00      2.00        0.00    1.00  0.00   
4      2.00  ... 2148.00      2.00      2.00        0.00    1.00  0.00   
6      1.00  ... 3298.00      2.00      2.00        0.00    0.00  1.00   

   Cancer  Heart  Diabetes  Comorbidity  
0    0.00   0.00      0.00            1  
1    0.00   0.00      0.00            1  
2    0.00   0.00      1.00            1  
4    0.00   0.00      0.00            1  
6    0.00   0.00      1.00            1  

[5 rows x 25 columns]

In [22]:
df.describe()

_AGE_G   MARITAL      _SEX   INCOME2   _EDUCAG     _RACE  _METSTAT  \
count 659144.00 659144.00 659144.00 659144.00 659144.00 659144.00 659144.00   
mean       4.42      2.31      1.53     18.67      3.06      1.97      1.31   
std        1.59      1.72      0.50     29.69      0.97      2.21      0.46   
min        1.00      1.00      1.00      1.00      1.00      1.00      1.00   
25%        3.00      1.00      1.00      5.00      2.00      1.00      1.00   
50%        5.00      1.00      2.00      7.00      3.00      1.00      1.00   
75%        6.00      3.00      2.00      9.00      4.00      1.00      2.00   
max        6.00      9.00      2.00     99.00      9.00      9.00      2.00   

       CHECKUP1    BPMEDS  _TOTINDA  ...     _BMI5  CHCKDNY2  VETERAN3  \
count 659144.00 659144.00 659144.00  ... 659144.00 659144.00 659144.00   
mean       1.43      1.67      1.25  ...   2847.10      1.97      1.88   
std        1.04      0.47      0.51  ...    647.48      0.33      0.41   
min        1.00      1.00      1.00  ...   1200.00      1.00      1.00   
25%        1.00      1.00      1.00  ...   2412.00      2.00      2.00   
50%        1.00      2.00      1.00  ...   2741.00      2.00      2.00   
75%        1.00      2.00      1.00  ...   3162.00      2.00      2.00   
max        9.00      2.00      9.00  ...   9933.00      9.00      9.00   

       Depression    Asthma      COPD    Cancer     Heart  Diabetes  \
count   659144.00 659144.00 659144.00 659144.00 659144.00 659144.00   
mean         0.20      0.90      0.08      0.18      0.11      0.13   
std          0.40      0.29      0.27      0.38      0.32      0.34   
min          0.00      0.00      0.00      0.00      0.00      0.00   
25%          0.00      1.00      0.00      0.00      0.00      0.00   
50%          0.00      1.00      0.00      0.00      0.00      0.00   
75%          0.00      1.00      0.00      0.00      0.00      0.00   
max          1.00      1.00      1.00      1.00      1.00      1.00   

       Comorbidity  
count    659144.00  
mean          0.97  
std           0.18  
min           0.00  
25%           1.00  
50%           1.00  
75%           1.00  
max           1.00  

[8 rows x 25 columns]

In [23]:
df.shape

(659144, 25)

<font color= white>The final Data Frames have the following Response Columns<font color= red>

* Depression
* Asthma
* COPD
* Cancer
* Heart
* Diabetes
* Comorbidity

<font color=white>And the following Feature Space<font color= red>

* _AGE_G
* MARITAL
* _SEX
* INCOME2
* _EDUCAG
* _RACE
* _METSTAT
* CHECKUP1
* BPMEDS
* _TOTINDA
* _SMOKER3
* _DRNKWK1
* _FRUTSU1
* _VEGESU1
* FRNCHDA_
* _BMI5
* CHCKDNY2
* VETERAN3

<font color=white>The shapes of the different Dat Frames are as follows:<font color= red>

1. 2019: <br>
325245, 25<br><br>
2. 2021: <br>
333899, 25<br><br>
3. Merged: <br>
659144, 25<br><br>

## Analysis 1
Using both years, run exploratory data analysis using crosstabs, visuals, and basic frequency distributions to understand how chronic conditions are distribituted across geography and demography. You may also use the newly created comorbidity variables for this analysis. 

- Summarize salient features of the healthiest and least healthy states in the country.
- Discuss if you noticed any associations of the risk factors such as Age, Sex, Income, Education, Marital Status etc. with the level of comorbidities, while comparing the two years of data.

## Analysis 2
1. Using only 2021, use several classification algorithms to classify comorbility variables 1 and 2:
- Logistic Regression
- KNN
- RF
- Gradient Boosting
- XGBoost
- Catboost

2. Write a short report describing the performance metrics. 
3. In the end, choose one model each for the classification of both categorical variables

### Import Packages Used

In [24]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import RandomOverSampler
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
import xgboost as xgb
from catboost import CatBoostClassifier

### Train Test Split

In [25]:
# Get feature space and response
multi = ['Depression', 'Asthma', 'COPD', 'Cancer', 'Heart', 'Diabetes']
binary = ['Comorbidity']

# Feature space
X = df_21_trim.copy()
X.drop(binary, inplace=True, axis=1)
X.drop(multi, inplace=True, axis=1)
print('Feature space size: ', X.shape, '\n')

# Standarize
std = ['_DRNKWK1', '_FRUTSU1', '_VEGESU1', 'FRNCHDA_', '_BMI5']
scalar = StandardScaler()
X[std] = scalar.fit_transform(X[std])

# Response
y_binary = df_21_trim[binary].copy()
y_multi = df_21_trim.apply(lambda row: sum(row[multi]), axis=1)
print('Response size:')
print('Binary: ', y_binary.shape)
print('Multi: ', y_multi.shape, '\n')

# Train Test Split
# Since we have many rows, 10% test is enough.
# Responses are highly imbalanced, so we over sample for each response.
ros = RandomOverSampler(random_state=0)
X_resampled_binary, y_resampled_binary = ros.fit_resample(X, y_binary)
X_train_binary, X_test_binary, y_train_binary, y_test_binary = train_test_split(X_resampled_binary, y_resampled_binary, test_size=0.1, random_state=42)
X_train_binary.reset_index(inplace=True, drop=True)
X_test_binary.reset_index(inplace=True, drop=True)
y_train_binary.reset_index(inplace=True, drop=True)
y_test_binary.reset_index(inplace=True, drop=True)

ros = RandomOverSampler(random_state=0)
X_resampled_multi, y_resampled_multi = ros.fit_resample(X, y_multi)
X_train_multi, X_test_multi, y_train_multi, y_test_multi = train_test_split(X_resampled_multi, y_resampled_multi, test_size=0.1, random_state=42)
X_train_multi.reset_index(inplace=True, drop=True)
X_test_multi.reset_index(inplace=True, drop=True)
y_train_multi.reset_index(inplace=True, drop=True)
y_test_multi.reset_index(inplace=True, drop=True)
print('Train set for binary case', X_train_binary.shape)
print('Train set for multi case', X_train_multi.shape)


Feature space size:  (333899, 18) 

Response size:
Binary:  (333899, 1)
Multi:  (333899,) 

Train set for binary case (581113, 18)
Train set for multi case (1107225, 18)


### Logistic Regresison

#### Binary

In [115]:
# Configure estimator
lgr = LogisticRegression(random_state=42,
                         class_weight='balanced')
parameters = {
    'C': [1e-3, 1e-2, 1e-1, 1],
}

# Grid search for binary case
clf = GridSearchCV(lgr, param_grid=parameters)
clf.fit(X_train_binary, y_train_binary)
y_predict = clf.predict(X_test_binary)
print('For the binary case, classification report on test set is:')
print(classification_report(y_test_binary, y_predict))
print('Best parameters are: ', clf.best_params_)
print('Best score on train set is: ', clf.best_score_)
print(pd.crosstab(y_predict, y_test_binary.Comorbidity, rownames=['y_predict'], colnames=['y_test']))

For the binary case, classification report on test set is:
              precision    recall  f1-score   support

           0       0.61      0.61      0.61     32363
           1       0.61      0.61      0.61     32206

    accuracy                           0.61     64569
   macro avg       0.61      0.61      0.61     64569
weighted avg       0.61      0.61      0.61     64569

Best parameters are:  {'C': 1}
Best score on train set is:  0.6092343494889769
y_test         0      1
y_predict              
0          19673  12628
1          12690  19578


#### Multi

In [116]:
# Configure estimator
lgr = LogisticRegression(random_state=42,
                         class_weight='balanced')
parameters = {
    'C': [1e-3, 1e-2, 1e-1, 1],
}

# Grid search for multi response
clf = GridSearchCV(lgr, param_grid=parameters)
clf.fit(X_train_multi, y_train_multi)
y_predict = clf.predict(X_test_multi)
print('For the multi response, classification report on test set is:')
print(classification_report(y_test_multi, y_predict))
print('Best parameters are: ', clf.best_params_)
print('Best score on train set is: ', clf.best_score_)
print(pd.crosstab(y_predict, y_test_multi, rownames=['y_predict'], colnames=['y_test']))

For the multi response, classification report on test set is:
              precision    recall  f1-score   support

         0.0       0.38      0.43      0.40     17896
         1.0       0.38      0.34      0.36     17641
         2.0       0.21      0.16      0.18     17445
         3.0       0.19      0.22      0.20     17475
         4.0       0.22      0.15      0.18     17440
         5.0       0.29      0.14      0.19     17583
         6.0       0.31      0.59      0.41     17545

    accuracy                           0.29    123025
   macro avg       0.28      0.29      0.27    123025
weighted avg       0.28      0.29      0.27    123025

Best parameters are:  {'C': 1}
Best score on train set is:  0.2884318905371537
y_test     0.00  1.00  2.00  3.00  4.00  5.00   6.00
y_predict                                           
0.00       7671  5244  3987  1889   851   507     99
1.00       4414  6083  2984  1319   767   510     98
2.00       2212  2272  2774  2512  1589  1147    6

### KNN

#### Binary

In [123]:
# Configure estimator
knn = KNeighborsClassifier(n_neighbors=5, leaf_size=10)

# Train and test binary case
knn.fit(X_train_binary, y_train_binary)
y_predict = knn.predict(X_test_binary)
print('For the binary case, classification report on test set is:')
print(classification_report(y_test_binary, y_predict))
print(pd.crosstab(y_predict, y_test_binary.Comorbidity, rownames=['y_predict'], colnames=['y_test']))


For the binary case, classification report on test set is:
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     32363
           1       1.00      0.90      0.95     32206

    accuracy                           0.95     64569
   macro avg       0.95      0.95      0.95     64569
weighted avg       0.95      0.95      0.95     64569

y_test         0      1
y_predict              
0          32363   3316
1              0  28890


#### Multi

In [124]:
# Configure estimator
knn = KNeighborsClassifier(n_neighbors=5, leaf_size=10)


# Train and test for multi response
knn.fit(X_train_multi, y_train_multi)
y_predict = knn.predict(X_test_multi)
print('For the multi response, classification report on test set is:')
print(classification_report(y_test_multi, y_predict))
print(pd.crosstab(y_predict, y_test_multi, rownames=['y_predict'], colnames=['y_test']))

For the multi response, classification report on test set is:
              precision    recall  f1-score   support

         0.0       0.85      1.00      0.92     17896
         1.0       0.63      0.45      0.53     17641
         2.0       0.61      0.48      0.54     17445
         3.0       0.75      0.90      0.81     17475
         4.0       0.91      1.00      0.95     17440
         5.0       0.98      1.00      0.99     17583
         6.0       1.00      1.00      1.00     17545

    accuracy                           0.83    123025
   macro avg       0.82      0.83      0.82    123025
weighted avg       0.82      0.83      0.82    123025

y_test      0.00  1.00  2.00   3.00   4.00   5.00   6.00
y_predict                                               
0.00       17896  1942  1160    137      0      0      0
1.00           0  7920  4047    547      0      0      0
2.00           0  4553  8385    908      3      0      0
3.00           0  2493  2792  15641      0      0      0

### Random Forest

#### Binary

In [125]:
# Configure estimator
rf = RandomForestClassifier(n_jobs=5)

# Train and test for binary response
rf.fit(X_train_binary, y_train_binary)
y_predict = rf.predict(X_test_binary)
print('For the binary response, classification report on test set is:')
print(classification_report(y_test_binary, y_predict))
print(pd.crosstab(y_predict, y_test_binary.Comorbidity, rownames=['y_predict'], colnames=['y_test']))

For the binary response, classification report on test set is:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     32363
           1       1.00      1.00      1.00     32206

    accuracy                           1.00     64569
   macro avg       1.00      1.00      1.00     64569
weighted avg       1.00      1.00      1.00     64569

y_test         0      1
y_predict              
0          32363      2
1              0  32204


#### Multi

In [126]:
# Configure estimator
rf = RandomForestClassifier(n_jobs=5)

# Train and test for multi response
rf.fit(X_train_multi, y_train_multi)
y_predict = rf.predict(X_test_multi)
print('For the multi response, classification report on test set is:')
print(classification_report(y_test_multi, y_predict))
print(pd.crosstab(y_predict, y_test_multi, rownames=['y_predict'], colnames=['y_test']))

For the multi response, classification report on test set is:
              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00     17896
         1.0       0.83      0.75      0.79     17641
         2.0       0.79      0.82      0.80     17445
         3.0       0.94      0.99      0.97     17475
         4.0       1.00      1.00      1.00     17440
         5.0       1.00      1.00      1.00     17583
         6.0       1.00      1.00      1.00     17545

    accuracy                           0.94    123025
   macro avg       0.94      0.94      0.94    123025
weighted avg       0.94      0.94      0.94    123025

y_test      0.00   1.00   2.00   3.00   4.00   5.00   6.00
y_predict                                                 
0.00       17896     20      3      0      0      0      0
1.00           0  13204   2680     37      0      0      0
2.00           0   3777  14274     63      0      0      0
3.00           0    604    450  17368      0  

### Gradient Boost

#### Binary

In [127]:
# Configure estimator
gbc = GradientBoostingClassifier()

# Train and test for binary response
gbc.fit(X_train_binary, y_train_binary)
y_predict = gbc.predict(X_test_binary)
print('For the binary response, classification report on test set is:')
print(classification_report(y_test_binary, y_predict))
print(pd.crosstab(y_predict, y_test_binary.Comorbidity, rownames=['y_predict'], colnames=['y_test']))

For the binary response, classification report on test set is:
              precision    recall  f1-score   support

           0       0.62      0.68      0.64     32363
           1       0.64      0.58      0.61     32206

    accuracy                           0.63     64569
   macro avg       0.63      0.63      0.63     64569
weighted avg       0.63      0.63      0.63     64569

y_test         0      1
y_predict              
0          21880  13636
1          10483  18570


#### Multi

In [128]:
# Configure estimator
gbc = GradientBoostingClassifier()

# Train and test for multi response
gbc.fit(X_train_multi, y_train_multi)
y_predict = gbc.predict(X_test_multi)
print('For the multi response, classification report on test set is:')
print(classification_report(y_test_multi, y_predict))
print(pd.crosstab(y_predict, y_test_multi, rownames=['y_predict'], colnames=['y_test']))

For the multi response, classification report on test set is:
              precision    recall  f1-score   support

         0.0       0.40      0.47      0.43     17896
         1.0       0.40      0.35      0.37     17641
         2.0       0.25      0.21      0.23     17445
         3.0       0.24      0.18      0.21     17475
         4.0       0.31      0.23      0.26     17440
         5.0       0.45      0.50      0.47     17583
         6.0       0.59      0.88      0.70     17545

    accuracy                           0.40    123025
   macro avg       0.38      0.40      0.38    123025
weighted avg       0.38      0.40      0.38    123025

y_test     0.00  1.00  2.00  3.00  4.00  5.00   6.00
y_predict                                           
0.00       8362  5716  4174  1805   633   258      0
1.00       4125  6117  2825  1242   621   215    192
2.00       2362  2552  3648  2949  1730   991    535
3.00       1313  1394  2401  3219  2556  1813    536
4.00        746   832  

### XGBoost

#### Binary

In [129]:
# Configure estimator
xbc = xgb.XGBClassifier(n_estimators=100,
						max_depth=10,
						eta=0.01,
						min_child_weight=5,
						random_state=100)

# Train and test for binary response
xbc.fit(X_train_binary, y_train_binary)
y_predict = xbc.predict(X_test_binary)
print('For the binary response, classification report on test set is:')
print(classification_report(y_test_binary, y_predict))
print(pd.crosstab(y_predict, y_test_binary.Comorbidity, rownames=['y_predict'], colnames=['y_test']))

For the binary response, classification report on test set is:
              precision    recall  f1-score   support

           0       0.67      0.79      0.73     32363
           1       0.75      0.61      0.67     32206

    accuracy                           0.70     64569
   macro avg       0.71      0.70      0.70     64569
weighted avg       0.71      0.70      0.70     64569

y_test         0      1
y_predict              
0          25720  12630
1           6643  19576


#### Multi

In [130]:
# Configure estimator
xbg = xgb.XGBClassifier(n_estimators=100,
						max_depth=10,
						eta=0.01,
						min_child_weight=5,
						random_state=100)

# Train and test for multi response
xbg.fit(X_train_multi, y_train_multi)
y_predict = xbg.predict(X_test_multi)
print('For the multi response, classification report on test set is:')
print(classification_report(y_test_multi, y_predict))
print(pd.crosstab(y_predict, y_test_multi, rownames=['y_predict'], colnames=['y_test']))

For the multi response, classification report on test set is:
              precision    recall  f1-score   support

         0.0       0.48      0.60      0.53     17896
         1.0       0.42      0.38      0.40     17641
         2.0       0.31      0.23      0.26     17445
         3.0       0.41      0.21      0.28     17475
         4.0       0.50      0.46      0.48     17440
         5.0       0.62      0.81      0.70     17583
         6.0       0.73      1.00      0.85     17545

    accuracy                           0.53    123025
   macro avg       0.50      0.53      0.50    123025
weighted avg       0.50      0.53      0.50    123025

y_test      0.00  1.00  2.00  3.00  4.00   5.00   6.00
y_predict                                             
0          10745  5445  3942  1723   551    131      0
1           3398  6645  3393  1532   573     95      0
2           1595  2522  4014  3026  1524    365      0
3            656   940  1598  3644  1541    537      0
4          

### Catboost

#### Binary

In [132]:
# Configure estimator
cbc = CatBoostClassifier(iterations=2,
                           depth=2,
                           learning_rate=1,
                           loss_function='Logloss',
                           verbose=True)

# Train and test for binary response
cbc.fit(X_train_binary, y_train_binary)
y_predict = cbc.predict(X_test_binary)
print('For the binary response, classification report on test set is:')
print(classification_report(y_test_binary, y_predict))
print(pd.crosstab(y_predict, y_test_binary.Comorbidity, rownames=['y_predict'], colnames=['y_test']))

0:	learn: 0.6711616	total: 27.9ms	remaining: 27.9ms
1:	learn: 0.6655812	total: 41ms	remaining: 0us
For the binary response, classification report on test set is:
              precision    recall  f1-score   support

           0       0.58      0.70      0.64     32363
           1       0.63      0.50      0.55     32206

    accuracy                           0.60     64569
   macro avg       0.61      0.60      0.60     64569
weighted avg       0.60      0.60      0.60     64569

y_test         0      1
y_predict              
0          22763  16174
1           9600  16032


#### Multi

In [ ]:
# Configure estimator
cbc = CatBoostClassifier(classes_count=7)

# Train and test for multi response
cbc.fit(X_train_multi, y_train_multi)
y_predict = cbc.predict(X_test_multi)
print('For the multi response, classification report on test set is:')
print(classification_report(y_test_multi, y_predict))
print(pd.crosstab(y_predict, y_test_multi, rownames=['y_predict'], colnames=['y_test']))

Learning rate set to 0.11313
0:	learn: 1.8906563	total: 682ms	remaining: 11m 20s
1:	learn: 1.8475926	total: 1.43s	remaining: 11m 52s
2:	learn: 1.8131224	total: 1.88s	remaining: 10m 24s
3:	learn: 1.7843307	total: 2.29s	remaining: 9m 30s
4:	learn: 1.7600900	total: 2.69s	remaining: 8m 56s
5:	learn: 1.7376300	total: 3.03s	remaining: 8m 22s
6:	learn: 1.7188428	total: 3.28s	remaining: 7m 45s
7:	learn: 1.7029984	total: 3.55s	remaining: 7m 20s
8:	learn: 1.6889682	total: 3.85s	remaining: 7m 3s
9:	learn: 1.6767446	total: 4.14s	remaining: 6m 49s
10:	learn: 1.6649258	total: 4.42s	remaining: 6m 37s
11:	learn: 1.6545250	total: 4.67s	remaining: 6m 24s
12:	learn: 1.6443511	total: 4.97s	remaining: 6m 17s
13:	learn: 1.6360854	total: 5.22s	remaining: 6m 7s
14:	learn: 1.6271393	total: 5.52s	remaining: 6m 2s
15:	learn: 1.6193288	total: 5.78s	remaining: 5m 55s
16:	learn: 1.6131532	total: 6.06s	remaining: 5m 50s
17:	learn: 1.6074141	total: 6.31s	remaining: 5m 44s
18:	learn: 1.6015096	total: 6.56s	remaining: 

### Analysis 2 Report

At first, I was trying to use a single train set and test set for all those chronic conditions and all models. But when I finished running those models, the performance on each condition is really poor that the model tend to predict most samples into one class. The multiclass case is even worse that actually all test cases are classified as 0 for all models.

The reason is that the classes are highly imbalanced for each chronic condition since only a small population has those conditions. Moreover, the population who have each condition is different from each other. So I over sampled for each condition, and over sampled again for the multiclass case. The resulting performance is then improved dramatically.

There are 5 features with a large range, `'_DRNKWK1', '_FRUTSU1', '_VEGESU1', 'FRNCHDA_', '_BMI5'`, that may affect the performance. So standardization is applied to those features. All other features have small range with 10, all they are categorical, so the rest are kept unchanged. The resulting performance is then slightly improved.

The table bellow shows the performance of each model on both responses. Scores are accuracy score calculated on the test set.

| Model               | Binary | Multiclass |
|---------------------|--------|------------|
| Logistic Regression | 0.61   | 0.29       |
| KNN                 | 0.95   | 0.83       |
| Random Forest       | 1.00   | 0.94       |
| Gradient Boost      | 0.63   | 0.40       |
| XGBoost             | 0.70   | 0.53       |
| CatBoost            | 0.60   | 0.59        |


Among those models random forest would be chosen as the model for classification because of its impressive performance and its speed in training. KNN could also be used for classify the binary category as it has a relatively high score.